lib/auth/apiserver.go

```diff
@@ -1586,23 +1586,29 @@ type samlAuthRawResponse struct {
 	HostSigners []json.RawMessage `json:"host_signers"`
 	// TLSCert is TLS certificate authority certificate
 	TLSCert []byte `json:"tls_cert,omitempty"`
+	// DiagInfo contains SSO Diagnostic Info
+	DiagInfo *SsoDiagnosticInfo `json:"diagInfo,omitempty"`
 }
 
 func (s *APIServer) validateSAMLResponse(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
 	var req *validateSAMLResponseReq
 	if err := httplib.ReadJSON(r, &req); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	response, err := auth.ValidateSAMLResponse(req.Response)
+
+	// we attach the diagnostic information either to the error or response.
+	// this is an unfortunate redundancy caused by limitation of our HTTP API
+	response, di, err := auth.ValidateSAMLResponse(req.Response)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, trace.Wrap(err).AddField("sso-diag-info", di)
```

Comment:
Can `di` field can be propagated by the `auth.ValidateSAMLResponse` handler. So the `auth.ValidateSAMLResponse` will return only `response and err` where err will have preset diagnostic ErrorField